File: src/com/almasb/zeph/entity/orion/Armor.java
package com.almasb.zeph.entity.orion;

import com.almasb.fxgl.entity.Entity;
import com.almasb.zeph.entity.orion.GameCharacter.Stat;

public class Armor extends EquippableItem {

    private static final long serialVersionUID = 826666654020818850L;

    public enum ArmorType {
        HELM, BODY, SHOES
    }

    public final ArmorType type;

    private int armor, marmor;

    public Armor(int id, String name, String description, String textureName,
            ItemLevel level, Element element, ArmorType type, int armor, int marmor, Rune... defRunes) {
        super(id, name, description, textureName, level, element, defRunes);
        this.type = type;
        this.armor = armor;
        this.marmor = marmor;
    }

    public Armor(Armor copy) {
        this(copy.id, copy.name, copy.description, copy.textureName,
                copy.level, copy.element, copy.type, copy.armor, copy.marmor, copy.defaultRunes.toArray(new Rune[0]));
    }

    @Override
    public void onEquip(Player ch) {
        super.onEquip(ch);
        ch.addBonusStat(Stat.ARM, getArmorRating());
        ch.addBonusStat(Stat.MARM, getMArmorRating());
    }

    @Override
    public void onUnEquip(Player ch) {
        super.onUnEquip(ch);
        ch.addBonusStat(Stat.ARM, -getArmorRating());
        ch.addBonusStat(Stat.MARM, -getMArmorRating());
    }

    public int getArmorRating() {
        return armor + refineLevel * (refineLevel > 2 ? level.bonus + 1 : level.bonus);
    }

    public int getMArmorRating() {
        return marmor + refineLevel * (refineLevel > 2 ? level.bonus + 1 : level.bonus);
    }

    @Override
    public Entity toEntity() {
        Entity e = new Entity("armor");
        e.setProperty("armor_data", this);
        return e;
    }

    public static class ArmorBuilder {

        public Armor build() {
            return null;
        }
    }
}

File: src/main/kotlin/com/almasb/zeph/item/Armor.kt
package com.almasb.zeph.item

import com.almasb.fxgl.dsl.getUIFactoryService
import com.almasb.zeph.character.CharacterEntity
import com.almasb.zeph.combat.Element
import com.almasb.zeph.combat.Stat
import javafx.beans.binding.Bindings
import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.scene.paint.Color
import java.util.concurrent.Callable

enum class ArmorType {
    HELM, BODY, SHOES
}

/**
 * Armor item.
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
class Armor(data: ArmorData) : EquipItem(data.description, data.itemLevel, data.runes, data.essences) {

    val element = SimpleObjectProperty<Element>()

    val armor = SimpleIntegerProperty()
    val marmor = SimpleIntegerProperty()

    val type = data.armorType

    init {
        element.value = data.element

        armor.bind(refineLevel.multiply(Bindings
                .`when`(refineLevel.greaterThan(2))
                .then(data.itemLevel.bonus + 1)
                .otherwise(data.itemLevel.bonus))
                .add(data.armor))

        marmor.bind(refineLevel.multiply(Bindings
                .`when`(refineLevel.greaterThan(2))
                .then(data.itemLevel.bonus + 1)
                .otherwise(data.itemLevel.bonus))
                .add(data.marmor))

        dynamicDescription.bind(SimpleStringProperty("")
                .concat(description.name + "\n")
                .concat(description.description + "\n")
                .concat(element.asString("Element: %s").concat("\n"))
                .concat(armor.asString("Armor: %d").concat("%\n"))
                .concat(marmor.asString("MArmor: %d").concat("%\n"))
                .concat(runes)
                .concat(essences)
        )

        dynamicTextFlow.children.addAll(
                getUIFactoryService().newText("Element: ", Color.WHITE, 14.0),
                getUIFactoryService().newText("", 16.0).also {
                    it.fillProperty().bind(Bindings.createObjectBinding(Callable { element.value.color }, element))
                    it.textProperty().bind(element.asString("%s\n"))
                },
                getUIFactoryService().newText("Armor: ", Color.WHITE, 14.0),
                getUIFactoryService().newText("", Color.GREEN, 16.0).also {
                    it.textProperty().bind(armor.asString("%d\n"))
                },
                getUIFactoryService().newText("MArmor: ", Color.WHITE, 14.0),
                getUIFactoryService().newText("", Color.BLUE, 16.0).also {
                    it.textProperty().bind(marmor.asString("%d\n"))
                }
        )
    }

    override fun onEquip(char: CharacterEntity) {
        super.onEquip(char)

        char.addBonus(Stat.ARM, armor.value)
        char.addBonus(Stat.MARM, marmor.value)
    }

    override fun onUnEquip(char: CharacterEntity) {
        super.onUnEquip(char)

        char.addBonus(Stat.ARM, -armor.value)
        char.addBonus(Stat.MARM, -marmor.value)
    }
}